cdn/signedurls/signurl.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Command signedurls contains helper function to sign URLs.
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+)
+
+// [START example]
+
+// SignURL creates a signed URL for an endpoint on Cloud CDN. sourceURL must
+// start with "https://" and should not have the Expires, KeyName, or Signature
+// query parameters. key should be in raw form (not base64url-encoded).
+func SignURL(sourceURL, keyName string, key []byte, expiration time.Time) string {
+	u := sourceURL
+	sep := "?"
+	if strings.Contains(u, "?") {
+		sep = "&"
+	}
+	u += sep
+	u += fmt.Sprintf("Expires=%d", expiration.Unix())
+	u += fmt.Sprintf("&KeyName=%s", keyName)
```

Comment:
Use url.Values and Encode

Reply:
This won't work because ordering here matters and net/url functions don't preserve.

Reply:
Should be fine for everything except appending `&Signature` to the end?

Reply:
Actually, what you have is good. Disregard.

Reply:
Yeah it still doesn't preserve the order when I parse existing URL and add my values.